contributors/design-proposals/service_links.md

```diff
@@ -0,0 +1,254 @@
+
+#### Pro/con: Simple API
+
+Pros:
+
+1.  Very small practical and conceptual delta from current behavior
+2.  Users can employ the same strategies for custom env vars
+3.  Easiest to explain
+4.  Easiest to support
+
+Cons:
+
+1.  Inflexible
+2.  Users have the same usability issues when custom env var names are required
+3.  Does not extend well to services in other namespaces
+
+#### Pro/con: Most complex API
+
+Pros:
+
+1.  Offers highest degree of flexibility
+2.  Custom env var names are cheap at fine-grained level of control
+
+Cons:
+
+1.  Most difficult to explain and reason about
+2.  Most difficult to support
+3.  Higher implementation cost
+4.  Large conceptual delta
+5.  Likely to add significant complexity to pod spec
+6.  Likely to be very verbose
+
+#### Lessons Learned from Prior APIs
+
+One fundamental lesson of Kubernetes API design is that we must make backward
+compatible API changes except in cases of special exception.  This means that we
+need to consider backward compatibility when making new API changes.  As a
+requirement, backward compatibility imposes a burden that you have to live with
+the APIs you release.  This means that we should attempt to be conservative with
+new API features and avoid unnecessary complexity.  Therefore, a simpler API is
+advantageous because complexity can be added later **if desired and warranted**.
+
+#### Middleground
+
+A conservative middleground might be an API that has behavior extremely close to
+the current implicit behavior but allows specifying custom variable names.  This
+approach has the advantage of being close to what users already expect, and can
+be implemented with a fairly simple API.
+
+### Non-existent Services
+
+When a service referenced using this API does not exist, the Kubelet should
+avoid starting the pod.
+
+## Proposed Changes
+
+### API changes
+
+#### Expressing dependencies on services
+
+```go
+type PodSpec {
+  // other fields omitted
+
+  // ServiceDependencies is a list of services in this pod's namespace that the
+  // pod depends on in order to start.  
+  ServiceDependencies []LocalObjectReference `json:"serviceDependencies,omitempty"`
+}
+```
+
+#### Environment variable injection
+
+Pull request [#148](https://github.com/kubernetes/community/pull/148)
+proposes an `EnvFrom` addition to the PodSpec to handle creating multiple
+environment variables from a single subject.  This API surface is ideal for the
+service use-cases, so we will extend it here:
+
+```go
+type Container struct {
+	// other fields omitted
+
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty"`
+}
+
+type EnvVarFromSource struct {
+	// other fields omitted
+
+	// Prefix is an optional prefix for generated environment variables
+	Prefix string `json:"prefix,omitempty"`
+
+	// Can be refactored to ObjectReference at a later time while
+	// maintaining backward compatibility.
+	Service *LocalObjectReference `json:"service,omitempty"`
```

Comment:
I am assuming this will follow the current convention.
Service will now be ServiceRef and its type is a *ServiceEnvSource.

Given explicit links above which do not inject but are required for the pod to start, does it also hold for ServiceRefs? Just trying to understand the relationship between the explicit links and ServiceRefs. What happens if you have a ServiceRef with no ServiceDependency?